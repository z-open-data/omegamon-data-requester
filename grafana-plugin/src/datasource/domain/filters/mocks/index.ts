export { TRANSFORM_FILTER_TEST_DATA } from './mockedFilterData';
