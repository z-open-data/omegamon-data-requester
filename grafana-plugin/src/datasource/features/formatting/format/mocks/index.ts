export { ColumnMetadataAndValues } from './ColumnMetadataAndValues';
export { numberEnumColumnMetadataAndValues, stringEnumColumnMetadataAndValues } from './enum';
export {
  columnMdsWithPrintf,
  columnMdsWithoutHexPrintf,
  numberCasesAsNumbersAndAsStrings,
  numberColumnMds,
  brokenNumberCasesAsStrings,
  numberCasesAsNumbers,
} from './number';
export { stringColumnMetadata, stringColumnValues } from './string';
export { invalidItmTimestamps, itmTimestampTestValues, timestampColumnMd, validItmTimestamps } from './timestamp';
