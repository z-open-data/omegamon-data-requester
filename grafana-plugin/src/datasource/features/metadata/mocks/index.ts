export { mockedAgents } from './mockedAgents';
export { mockedApplicationMetadatas } from './mockedApplicationMetadatas';
export { mockedMetadataLoader } from './mockedMetadataLoader';
export { mockedTableMetadatas, mockedTablesResponse } from './mockedTableMetadatas';
