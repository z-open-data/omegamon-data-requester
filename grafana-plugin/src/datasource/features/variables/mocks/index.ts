export * from './metricsQueryFilters';
export * from './mockedTableMetadatas';
export * from './replaceFnInputOutputMap';
export * from './scopedVars';
export * from './variableModels';
