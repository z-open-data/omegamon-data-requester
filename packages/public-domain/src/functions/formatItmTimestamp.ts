const INVALID_TIMESTAMP = new Date(NaN);

const ITM_TIMESTAMP_FORMAT = 'CYYMMDDhhmmss000';
const ITM_TIMESTAMP_LENGTH = ITM_TIMESTAMP_FORMAT.length;

function validateItmTimestamp(itmTimestamp: string): boolean {
  const hasOnlyDigits = /^\d+$/.test(itmTimestamp);
  const isValidLength = itmTimestamp.length === ITM_TIMESTAMP_LENGTH;

  return (
    hasOnlyDigits &&
    isValidLength &&
    /**
     * See features/format/README.md
     */
    itmTimestamp !== '0000000000000000'
  );
}

/**
 * Caution! This function treats itmTimestamp as if it
 * defined time in the same time zone, where the code is running.
 * In some cases that does not matter,
 * but in other cases it might lead to bugs.
 */
function itmTimestampToUnixTimestamp(itmTimestamp: string): Date {
  const century = Number(itmTimestamp[0]) + 19;
  const year = Number(itmTimestamp.substring(1, 3));
  const month = Number(itmTimestamp.substring(3, 5)) - 1;
  const day = Number(itmTimestamp.substring(5, 7));
  const hour = Number(itmTimestamp.substring(7, 9));
  const minute = Number(itmTimestamp.substring(9, 11));
  const second = Number(itmTimestamp.substring(11, 13));
  const millisecond = Number(itmTimestamp.substring(13, 16));

  const date = new Date(century * 100 + year, month, day, hour, minute, second, millisecond);
  return date;
}

/**
 * @param itmTimestamp - ITM timestamp (possibly corrupted)
 * @returns UNIX timestamp or NaN (if ITM timestamp is corrupted)
 */
export function formatItmTimestamp(itmTimestamp: string | number): Date {
  const trimmed = `${itmTimestamp}`.trim();

  if (!validateItmTimestamp(trimmed)) {
    return INVALID_TIMESTAMP;
  }

  const date = itmTimestampToUnixTimestamp(trimmed);
  return date;
}
