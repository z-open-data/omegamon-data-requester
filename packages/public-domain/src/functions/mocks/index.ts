export { invalidItmTimestamps, itmTimestampTestValues, timestampColumnMd, validItmTimestamps } from './timestamp';
