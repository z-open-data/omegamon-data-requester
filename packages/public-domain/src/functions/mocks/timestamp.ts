// eslint-disable-next-line no-restricted-imports
import { ColumnMetadata } from '../../types';

export const validItmTimestamps = [
  '0230608055952216', // 20th century
  '1230608055952216',
  '1230608044722135',
  '1230608044722133',
  '1230608044722131',
  '1230608044722196',
  '1230603095232971',
  '   1230608054618110  ', // untrimmed
  '2230608054618110', // 22nd century lol
];

export const invalidItmTimestamps = [
  /**
   * See features/format/README.md
   */
  '0000000000000000', // special case
  '230608044723635', // missing century digit as first char
  '123060abcde28656', // letters
  '12306;/-95232971', // symbols
  '123060804472213398', // too long
  '12306030952', // too short
  '', // empty string
];

export const itmTimestampTestValues = [...validItmTimestamps, ...invalidItmTimestamps];

export const timestampColumnMd = {
  attributeName: 'Application_Start_Time',
  defaultSortDirection: 'A',
  description: 'The time and date when this VTAM application was started.',
  extensions: {
    COST: '0',
    NULL: 'N',
    SHARE: 'Y',
    USAGE: 'N',
  },
  id: 'STARTTIME',
  maxLength: 16,
  name: 'Application\nStart\nTime',
  odpName: '',
  sortType: 'lexical',
  type: 'itmTimestamp',
  version: 0,
  atomize: false,
} satisfies ColumnMetadata;
