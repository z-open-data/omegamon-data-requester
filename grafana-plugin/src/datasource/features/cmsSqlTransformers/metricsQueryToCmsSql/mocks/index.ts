import { ColumnMetadata, ColumnMetadataMap, TableMetadata, AffinityId } from 'public-domain';

import { mockedApplicationMetadatas } from 'datasource/features/metadata/mocks';

import { MetricsQueryToCmsSqlTests } from './types';

const inodestsMetadata = {
  applicationCode: 'O4SRV',
  id: 'INODESTS',
  columns: {
    AFFINITIES: { id: 'AFFINITIES', type: 'string' } as ColumnMetadata,
    EXPIRYTIME: { id: 'EXPIRYTIME', type: 'number' } as ColumnMetadata,
  } as ColumnMetadataMap,
} as TableMetadata;

const realthdaMetadata = {
  applicationCode: 'KDP',
  id: 'REALTHDA',
  columns: {
    ORIGINNODE: { id: 'ORIGINNODE', type: 'string' } as ColumnMetadata,
  } as ColumnMetadataMap,
} as TableMetadata;

const db2connMetadata = {
  applicationCode: 'KDP',
  id: 'DB2CONN',
  columns: {
    EESGD028: { id: 'EESGD028', type: 'number' } as ColumnMetadata,
  } as ColumnMetadataMap,
} as TableMetadata;

const reqularQueryTests: MetricsQueryToCmsSqlTests = {
  validMetricsQueries: [
    {
      name: 'SELECT FROM clause',
      queryParams: {
        affinityId: '%IBM.ITM_INTERNAL' as AffinityId,
        tableId: 'INODESTS',
        columns: [
          {
            id: 'AFFINITIES',
          },
        ],
        agentsAndGroups: [],
        filter: {},
        history: false,
        orderBy: [],
        groupBy: [],
        parmas: [],
        first: 0,
      },
      applications: mockedApplicationMetadatas,
      metadata: inodestsMetadata,
      expectedOutput: 'SELECT AFFINITIES FROM O4SRV.INODESTS;',
    },
    {
      name: 'multiple SELECT FROM clause',
      queryParams: {
        affinityId: '%IBM.ITM_INTERNAL' as AffinityId,
        tableId: 'INODESTS',
        columns: [
          {
            id: 'AFFINITIES',
          },
          {
            id: 'EXPIRYTIME',
          },
        ],
        agentsAndGroups: [],
        filter: {},
        history: false,
        orderBy: [],
        groupBy: [],
        parmas: [],
        first: 0,
      },
      applications: mockedApplicationMetadatas,
      metadata: inodestsMetadata,
      expectedOutput: 'SELECT AFFINITIES, EXPIRYTIME FROM O4SRV.INODESTS;',
    },
    {
      name: "WHERE clause with SYSTEM.PARMA('NODELIST') array",
      queryParams: {
        affinityId: '%IBM.STATIC017' as AffinityId,
        tableId: 'REALTHDA',
        columns: [
          {
            id: 'ORIGINNODE',
          },
        ],
        agentsAndGroups: [
          { id: 'agentOrGroupName', name: 'IC1A:RSD1:DB2' },
          { id: 'agentOrGroupName', name: 'ICA4:RSD1:DB2' },
        ],
        filter: {},
        history: false,
        orderBy: [],
        groupBy: [],
        parmas: [],
        first: 0,
      },
      metadata: realthdaMetadata,
      applications: mockedApplicationMetadatas,
      expectedOutput:
        "SELECT ORIGINNODE FROM KDP.REALTHDA WHERE SYSTEM.PARMA('NODELIST', 'IC1A:RSD1:DB2, ICA4:RSD1:DB2', 28);",
    },
    {
      name: 'Aggregation function',
      queryParams: {
        affinityId: '%IBM.ITM_INTERNAL' as AffinityId,
        tableId: 'INODESTS',
        columns: [
          {
            id: 'AFFINITIES',
            aggregationFunction: 'MAX',
          },
        ],
        agentsAndGroups: [],
        filter: {},
        history: false,
        orderBy: [],
        groupBy: [],
        parmas: [],
        first: 0,
      },
      applications: mockedApplicationMetadatas,
      metadata: inodestsMetadata,
      expectedOutput: 'SELECT MAX(AFFINITIES) FROM O4SRV.INODESTS;',
    },
    {
      name: 'HISTORY',
      queryParams: {
        affinityId: '%IBM.ITM_INTERNAL' as AffinityId,
        tableId: 'INODESTS',
        columns: [
          {
            id: 'AFFINITIES',
          },
        ],
        history: true,
        agentsAndGroups: [],
        filter: {},
        orderBy: [],
        groupBy: [],
        parmas: [],
        first: 0,
      },
      applications: mockedApplicationMetadatas,
      metadata: inodestsMetadata,
      expectedOutput: 'SELECT AFFINITIES FROM O4SRV.INODESTS HISTORY();',
    },
    {
      name: 'ORDER BY',
      queryParams: {
        affinityId: '%IBM.ITM_INTERNAL' as AffinityId,
        tableId: 'INODESTS',
        columns: [
          {
            id: 'AFFINITIES',
          },
        ],
        orderBy: [
          {
            columnId: 'AFFINITIES',
            type: 'DESC',
          },
        ],
        agentsAndGroups: [],
        filter: {},
        history: false,
        groupBy: [],
        parmas: [],
        first: 0,
      },
      applications: mockedApplicationMetadatas,
      metadata: inodestsMetadata,
      expectedOutput: 'SELECT AFFINITIES FROM O4SRV.INODESTS ORDER BY AFFINITIES DESC;',
    },
    {
      name: 'Multiple ORDER BY',
      queryParams: {
        affinityId: '%IBM.ITM_INTERNAL' as AffinityId,
        tableId: 'INODESTS',
        columns: [
          {
            id: 'AFFINITIES',
          },
          {
            id: 'EXPIRYTIME',
          },
        ],
        orderBy: [
          {
            columnId: 'AFFINITIES',
            type: 'DESC',
          },
          {
            columnId: 'EXPIRYTIME',
            type: 'ASC',
          },
        ],
        agentsAndGroups: [],
        filter: {},
        history: false,
        groupBy: [],
        parmas: [],
        first: 0,
      },
      applications: mockedApplicationMetadatas,
      metadata: inodestsMetadata,
      expectedOutput: 'SELECT AFFINITIES, EXPIRYTIME FROM O4SRV.INODESTS ORDER BY AFFINITIES DESC, EXPIRYTIME ASC;',
    },
    {
      name: 'Multiple GROUP BY',
      queryParams: {
        affinityId: '%IBM.STATIC017' as AffinityId,
        tableId: 'REALTHDA',
        columns: [
          {
            id: 'ORIGINNODE',
          },
          {
            id: 'TIMESTAMP',
          },
        ],
        groupBy: ['ORIGINNODE', 'TIMESTAMP'],
        agentsAndGroups: [],
        filter: {},
        history: false,
        orderBy: [],
        parmas: [],
        first: 0,
      },
      applications: mockedApplicationMetadatas,
      metadata: realthdaMetadata,
      expectedOutput: 'SELECT ORIGINNODE, TIMESTAMP FROM KDP.REALTHDA GROUP BY ORIGINNODE, TIMESTAMP;',
    },
    {
      name: 'One parma',
      queryParams: {
        affinityId: '%IBM.STATIC017' as AffinityId,
        tableId: 'REALTHDA',
        columns: [
          {
            id: 'ORIGINNODE',
          },
        ],
        parmas: [
          {
            name: 'IGNOREFILTER',
            value: 'r',
            length: 1,
          },
        ],
        agentsAndGroups: [],
        filter: {},
        history: false,
        orderBy: [],
        groupBy: [],
        first: 0,
      },
      applications: mockedApplicationMetadatas,
      metadata: realthdaMetadata,
      expectedOutput: "SELECT ORIGINNODE FROM KDP.REALTHDA WHERE SYSTEM.PARMA('IGNOREFILTER', 'r', 1);",
    },
    {
      name: 'Multiple parmas',
      queryParams: {
        affinityId: '%IBM.STATIC017' as AffinityId,
        tableId: 'REALTHDA',
        columns: [
          {
            id: 'ORIGINNODE',
          },
        ],
        parmas: [
          {
            name: 'IGNOREFILTER',
            value: 'r',
            length: 1,
          },
          {
            name: 'ORIGINNODE',
            value: 'IZSME:LZ',
            length: 8,
          },
        ],
        agentsAndGroups: [],
        filter: {},
        history: false,
        orderBy: [],
        groupBy: [],
        first: 0,
      },
      applications: mockedApplicationMetadatas,
      metadata: realthdaMetadata,
      expectedOutput:
        "SELECT ORIGINNODE FROM KDP.REALTHDA WHERE SYSTEM.PARMA('IGNOREFILTER', 'r', 1) AND SYSTEM.PARMA('ORIGINNODE', 'IZSME:LZ', 8);",
    },
    {
      name: 'One parma with just name',
      queryParams: {
        affinityId: '%IBM.STATIC017' as AffinityId,
        tableId: 'REALTHDA',
        columns: [
          {
            id: 'ORIGINNODE',
          },
        ],
        parmas: [
          {
            name: 'ORIGINNODE',
          },
        ],
        agentsAndGroups: [],
        filter: {},
        history: false,
        orderBy: [],
        groupBy: [],
        first: 0,
      },
      applications: mockedApplicationMetadatas,
      metadata: realthdaMetadata,
      expectedOutput: "SELECT ORIGINNODE FROM KDP.REALTHDA WHERE SYSTEM.PARMA('ORIGINNODE');",
    },
    {
      name: 'Parmas without length',
      queryParams: {
        affinityId: '%IBM.STATIC017' as AffinityId,
        tableId: 'REALTHDA',
        columns: [
          {
            id: 'ORIGINNODE',
          },
        ],
        parmas: [
          {
            name: 'IGNOREFILTER',
            value: 'r',
          },
        ],
        agentsAndGroups: [],
        filter: {},
        history: false,
        orderBy: [],
        groupBy: [],
        first: 0,
      },
      applications: mockedApplicationMetadatas,
      metadata: realthdaMetadata,
      expectedOutput: "SELECT ORIGINNODE FROM KDP.REALTHDA WHERE SYSTEM.PARMA('IGNOREFILTER', 'r');",
    },
    {
      name: 'First',
      queryParams: {
        affinityId: '%IBM.STATIC017' as AffinityId,
        tableId: 'REALTHDA',
        columns: [
          {
            id: 'ORIGINNODE',
          },
        ],
        first: 5,
        agentsAndGroups: [],
        filter: {},
        history: false,
        orderBy: [],
        groupBy: [],
        parmas: [],
      },
      applications: mockedApplicationMetadatas,
      metadata: realthdaMetadata,
      expectedOutput: 'SELECT ORIGINNODE FROM KDP.REALTHDA WHERE FIRST(5);',
    },
    {
      name: 'single filter',
      queryParams: {
        affinityId: '%IBM.ITM_INTERNAL' as AffinityId,
        tableId: 'INODESTS',
        columns: [
          {
            id: 'AFFINITIES',
          },
        ],
        filter: {
          nonAggregated: {
            clause: {
              columnId: 'AFFINITIES',
              operator: '=',
              userDefinedValue: '%IBM.CICSplex 00000000050000yA043',
            },
          },
        },
        agentsAndGroups: [],
        history: false,
        orderBy: [],
        groupBy: [],
        parmas: [],
        first: 0,
      },
      applications: mockedApplicationMetadatas,
      metadata: inodestsMetadata,
      expectedOutput: "SELECT AFFINITIES FROM O4SRV.INODESTS WHERE AFFINITIES = '%IBM.CICSplex 00000000050000yA043';",
    },
    {
      name: 'Filter with numeric value',
      queryParams: {
        affinityId: '%IBM.STATIC017' as AffinityId,
        tableId: 'DB2CONN',
        columns: [
          {
            id: 'EESGD028',
          },
        ],
        filter: {
          nonAggregated: {
            clause: {
              columnId: 'EESGD028',
              operator: '>',
              userDefinedValue: '0',
            },
          },
        },
        agentsAndGroups: [],
        history: false,
        orderBy: [],
        groupBy: [],
        parmas: [],
        first: 0,
      },
      applications: mockedApplicationMetadatas,
      metadata: db2connMetadata,
      expectedOutput: 'SELECT EESGD028 FROM KDP.DB2CONN WHERE EESGD028 > 0;',
    },
    {
      name: 'Filter with variable value for column of "number" type',
      queryParams: {
        affinityId: '%IBM.STATIC017' as AffinityId,
        tableId: 'DB2CONN',
        columns: [
          {
            id: 'EESGD028',
          },
        ],
        filter: {
          nonAggregated: {
            clause: {
              columnId: 'EESGD028',
              operator: '>',
              userDefinedValue: '$VARIABLE',
            },
          },
        },
        agentsAndGroups: [],
        history: false,
        orderBy: [],
        groupBy: [],
        parmas: [],
        first: 0,
      },
      applications: mockedApplicationMetadatas,
      metadata: db2connMetadata,
      expectedOutput: 'SELECT EESGD028 FROM KDP.DB2CONN WHERE EESGD028 > $VARIABLE;',
    },
    {
      name: 'Filter with variable value for column of "string" type',
      queryParams: {
        affinityId: '%IBM.ITM_INTERNAL' as AffinityId,
        tableId: 'INODESTS',
        columns: [
          {
            id: 'AFFINITIES',
          },
        ],
        filter: {
          nonAggregated: {
            clause: {
              columnId: 'AFFINITIES',
              operator: '=',
              userDefinedValue: '$VARIABLE',
            },
          },
        },
        agentsAndGroups: [],
        history: false,
        orderBy: [],
        groupBy: [],
        parmas: [],
        first: 0,
      },
      applications: mockedApplicationMetadatas,
      metadata: inodestsMetadata,
      expectedOutput: "SELECT AFFINITIES FROM O4SRV.INODESTS WHERE AFFINITIES = '$VARIABLE';",
    },
    {
      name: 'OR filter',
      queryParams: {
        affinityId: '%IBM.ITM_INTERNAL' as AffinityId,
        tableId: 'INODESTS',
        columns: [
          {
            id: 'AFFINITIES',
          },
        ],
        filter: {
          nonAggregated: {
            or: [
              {
                clause: {
                  columnId: 'AFFINITIES',
                  operator: '=',
                  userDefinedValue: '%IBM.CICSplex 00000000050000yA043',
                },
              },
              {
                clause: {
                  columnId: 'AFFINITIES',
                  operator: '=',
                  userDefinedValue: '%IBM.CICSplex 00000000050000yA022',
                },
              },
            ],
          },
        },
        agentsAndGroups: [],
        history: false,
        orderBy: [],
        groupBy: [],
        parmas: [],
        first: 0,
      },
      applications: mockedApplicationMetadatas,
      metadata: inodestsMetadata,
      expectedOutput:
        "SELECT AFFINITIES FROM O4SRV.INODESTS WHERE (AFFINITIES = '%IBM.CICSplex 00000000050000yA043' OR AFFINITIES = '%IBM.CICSplex 00000000050000yA022');",
    },
    {
      name: 'AND filter',
      queryParams: {
        affinityId: '%IBM.ITM_INTERNAL' as AffinityId,
        tableId: 'INODESTS',
        columns: [
          {
            id: 'AFFINITIES',
          },
          {
            id: 'EXPIRYTIME',
          },
        ],
        filter: {
          nonAggregated: {
            and: [
              {
                clause: {
                  columnId: 'AFFINITIES',
                  operator: '=',
                  userDefinedValue: '%IBM.CICSplex 00000000050000yA043',
                },
              },
              {
                clause: {
                  columnId: 'EXPIRYTIME',
                  operator: '=',
                  userDefinedValue: '9',
                },
              },
            ],
          },
        },
        agentsAndGroups: [],
        history: false,
        orderBy: [],
        groupBy: [],
        parmas: [],
        first: 0,
      },
      applications: mockedApplicationMetadatas,
      metadata: inodestsMetadata,
      expectedOutput:
        "SELECT AFFINITIES, EXPIRYTIME FROM O4SRV.INODESTS WHERE (AFFINITIES = '%IBM.CICSplex 00000000050000yA043' AND EXPIRYTIME = 9);",
    },
    {
      name: 'OR filter inside AND',
      queryParams: {
        affinityId: '%IBM.ITM_INTERNAL' as AffinityId,
        tableId: 'INODESTS',
        columns: [
          {
            id: 'AFFINITIES',
          },
        ],
        filter: {
          nonAggregated: {
            and: [
              {
                clause: {
                  columnId: 'AFFINITIES',
                  operator: '=',
                  userDefinedValue: '%IBM.CICSplex 00000000050000yA043',
                },
              },
              {
                or: [
                  {
                    clause: {
                      columnId: 'AFFINITIES',
                      operator: '=',
                      userDefinedValue: '%IBM.CICSplex 00000000050000yA043',
                    },
                  },
                  {
                    clause: {
                      columnId: 'AFFINITIES',
                      operator: '=',
                      userDefinedValue: '%IBM.CICSplex 00000000050000yA033',
                    },
                  },
                ],
              },
            ],
          },
        },
        agentsAndGroups: [],
        history: false,
        orderBy: [],
        groupBy: [],
        parmas: [],
        first: 0,
      },
      applications: mockedApplicationMetadatas,
      metadata: inodestsMetadata,
      expectedOutput:
        "SELECT AFFINITIES FROM O4SRV.INODESTS WHERE (AFFINITIES = '%IBM.CICSplex 00000000050000yA043' AND (AFFINITIES = '%IBM.CICSplex 00000000050000yA043' OR AFFINITIES = '%IBM.CICSplex 00000000050000yA033'));",
    },
    {
      name: 'SYSTEM.PARMA Only nodelist',
      queryParams: {
        affinityId: '%IBM.ITM_INTERNAL' as AffinityId,
        tableId: 'INODESTS',
        columns: [
          {
            id: 'AFFINITIES',
          },
        ],
        filter: {},
        agentsAndGroups: [
          { id: 'agentOrGroupName', name: '*MVS_TEST' },
          { id: 'agentOrGroupName', name: 'SECOND_ON' },
        ],
        history: false,
        orderBy: [],
        groupBy: [],
        parmas: [],
        first: 0,
      },
      applications: mockedApplicationMetadatas,
      metadata: inodestsMetadata,
      expectedOutput:
        "SELECT AFFINITIES FROM O4SRV.INODESTS WHERE SYSTEM.PARMA('NODELIST', '*MVS_TEST, SECOND_ON', 20);",
    },
    {
      name: 'SYSTEM.PARMA Nodelist and standart parma',
      queryParams: {
        affinityId: '%IBM.ITM_INTERNAL' as AffinityId,
        tableId: 'INODESTS',
        columns: [
          {
            id: 'AFFINITIES',
          },
        ],
        filter: {},
        agentsAndGroups: [
          { id: 'agentOrGroupName', name: '*MVS_TEST' },
          { id: 'agentOrGroupName', name: 'SECOND_ON' },
        ],
        history: false,
        orderBy: [],
        groupBy: [],
        parmas: [
          {
            name: 'PARMA',
            value: 'Second',
            length: 6,
          },
        ],
        first: 0,
      },
      applications: mockedApplicationMetadatas,
      metadata: inodestsMetadata,
      expectedOutput:
        "SELECT AFFINITIES FROM O4SRV.INODESTS WHERE SYSTEM.PARMA('PARMA', 'Second', 6) AND SYSTEM.PARMA('NODELIST', '*MVS_TEST, SECOND_ON', 20);",
    },
  ],
  invalidMetricsQueries: [
    {
      name: 'applicationCode provided but no tableId',
      queryParams: {
        affinityId: '%IBM.STATIC017' as AffinityId,
        tableId: '',
        columns: [],
        filter: {},
        history: false,
        orderBy: [],
        groupBy: [],
        parmas: [],
        first: 0,
        agentsAndGroups: [],
      },
      applications: mockedApplicationMetadatas,
      metadata: realthdaMetadata,
      expectedOutput: 'SELECT  FROM KDP.;',
    },
    {
      name: 'applicationCode, agentsAndGroups but no tableId is provided',
      queryParams: {
        affinityId: '%IBM.STATIC017' as AffinityId,
        tableId: '',
        columns: [],
        agentsAndGroups: [
          { id: 'agentOrGroupName', name: '*MVS_TEST' },
          { id: 'agentOrGroupName', name: 'SECOND_ON' },
        ],
        history: false,
        filter: {},
        orderBy: [],
        groupBy: [],
        parmas: [],
        first: 0,
      },
      applications: mockedApplicationMetadatas,
      metadata: null,
      expectedOutput: "SELECT  FROM KDP. WHERE SYSTEM.PARMA('NODELIST', '*MVS_TEST, SECOND_ON', 20);",
    },
    {
      name: 'applicationCode and tableId but no columns is provided',
      queryParams: {
        affinityId: '%IBM.STATIC017' as AffinityId,
        tableId: 'REALTHDA',
        columns: [],
        agentsAndGroups: [],
        history: false,
        filter: {},
        orderBy: [],
        groupBy: [],
        parmas: [],
        first: 0,
      },
      applications: mockedApplicationMetadatas,
      metadata: realthdaMetadata,
      expectedOutput: 'SELECT  FROM KDP.REALTHDA;',
    },
    {
      name: 'no column found in table metadata',
      queryParams: {
        affinityId: '%IBM.STATIC017' as AffinityId,
        tableId: 'REALTHDA',
        columns: [
          {
            id: 'INCORRECT_ID',
          },
        ],
        agentsAndGroups: [],
        history: false,
        filter: {
          nonAggregated: {
            clause: {
              columnId: 'INCORRECT_ID',
              operator: '=',
              userDefinedValue: '$VARIABLE',
            },
          },
        },
        orderBy: [],
        groupBy: [],
        parmas: [],
        first: 0,
      },
      applications: mockedApplicationMetadatas,
      metadata: realthdaMetadata,
      expectedOutput: `SELECT INCORRECT_ID FROM KDP.REALTHDA WHERE INCORRECT_ID = '$VARIABLE';`,
    },
    {
      name: 'Filter clause input not convertable to number',
      queryParams: {
        affinityId: '%IBM.STATIC007' as AffinityId,
        tableId: 'ASCPUUTIL',
        columns: [
          {
            id: 'ORIGINNODE',
            aggregationFunction: 'COUNT',
          },
          {
            id: 'ASID',
          },
        ],
        agentsAndGroups: [
          {
            id: 'agentOrGroupName',
            name: 'RSPLUI01:RS25:MVSSYS',
          },
          {
            id: 'agentOrGroupName',
            name: '*MVS_SYSTEM',
          },
        ],
        filter: {
          nonAggregated: {
            clause: {
              columnId: 'Type',
              operator: '<>',
              userDefinedValue: '6dtgfd',
            },
          },
        },
        parmas: [],
        history: false,
        groupBy: ['ASID'],
        orderBy: [],
      },
      applications: mockedApplicationMetadatas,
      metadata: realthdaMetadata,
      expectedOutput: `SELECT COUNT(ORIGINNODE), ASID FROM KM5.ASCPUUTIL WHERE Type <> '6dtgfd' AND SYSTEM.PARMA('NODELIST', 'RSPLUI01:RS25:MVSSYS, *MVS_SYSTEM', 33) GROUP BY ASID;`,
    },
  ],
};

export default reqularQueryTests;
