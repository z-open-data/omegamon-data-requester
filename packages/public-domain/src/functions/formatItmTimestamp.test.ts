import { formatItmTimestamp } from './formatItmTimestamp';
import { validItmTimestamps, invalidItmTimestamps } from './mocks';

describe(formatItmTimestamp.name, () => {
  validItmTimestamps.forEach((itmTimestamp) => {
    test(`"${itmTimestamp}"`, () => {
      expect(formatItmTimestamp(itmTimestamp)).toMatchSnapshot();
    });
  });
  describe('invalid ITM timestamps', () => {
    invalidItmTimestamps.forEach((invalidItmTimestamp) => {
      test(invalidItmTimestamp, () => {
        const invalidDate = formatItmTimestamp(invalidItmTimestamp);
        expect(invalidDate.toString()).toEqual('Invalid Date');
      });
    });
  });
});
