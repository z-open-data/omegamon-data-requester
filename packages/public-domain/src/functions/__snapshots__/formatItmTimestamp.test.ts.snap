// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formatItmTimestamp "   1230608054618110  " 1`] = `2023-06-08T05:46:18.110Z`;

exports[`formatItmTimestamp "0230608055952216" 1`] = `1923-06-08T05:59:52.216Z`;

exports[`formatItmTimestamp "1230603095232971" 1`] = `2023-06-03T09:52:32.971Z`;

exports[`formatItmTimestamp "1230608044722131" 1`] = `2023-06-08T04:47:22.131Z`;

exports[`formatItmTimestamp "1230608044722133" 1`] = `2023-06-08T04:47:22.133Z`;

exports[`formatItmTimestamp "1230608044722135" 1`] = `2023-06-08T04:47:22.135Z`;

exports[`formatItmTimestamp "1230608044722196" 1`] = `2023-06-08T04:47:22.196Z`;

exports[`formatItmTimestamp "1230608055952216" 1`] = `2023-06-08T05:59:52.216Z`;

exports[`formatItmTimestamp "2230608054618110" 1`] = `2123-06-08T05:46:18.110Z`;
